frontend/src/metabase/admin/permissions/selectors/confirmations.ts

```diff
@@ -157,38 +158,69 @@ export function getWillRevokeNativeAccessWarningModal(
+export function getViewDataPermissionsTooRestrictiveWarningModal(
   permissions: GroupsPermissions,
   groupId: Group["id"],
-  entityId: EntityId,
+  entityId: DatabaseEntityId | SchemaEntityId,
+  database: Database,
   value: DataPermissionValue,
 ) {
-  const nativePermission = getSchemasPermission(
-    permissions,
-    groupId,
-    entityId,
-    DataPermission.CREATE_QUERIES,
-  );
+  // if user sets 'Query builder and native' for a DB, warn them that view data permissions must be 'Unrestricted'
+  if (!isSchemaEntityId(entityId)) {
+    const nativePermission = getSchemasPermission(
+      permissions,
+      groupId,
+      entityId,
+      DataPermission.CREATE_QUERIES,
+    );
 
-  const viewPermission = getSchemasPermission(
+    const viewPermission = getSchemasPermission(
+      permissions,
+      groupId,
+      entityId,
+      DataPermission.VIEW_DATA,
+    );
+
+    if (
+      value === DataPermissionValue.QUERY_BUILDER_AND_NATIVE &&
+      nativePermission !== DataPermissionValue.QUERY_BUILDER_AND_NATIVE &&
+      PLUGIN_ADVANCED_PERMISSIONS.shouldShowViewDataColumn &&
+      ![
+        DataPermissionValue.UNRESTRICTED,
+        DataPermissionValue.IMPERSONATED,
+      ].includes(viewPermission)
+    ) {
+      return {
+        title: t`Allow native query editing?`,
+        message: t`This will also change this group's data access to Unrestricted for this database.`,
+        confirmButtonText: t`Allow`,
+        cancelButtonText: t`Cancel`,
+      };
+    }
+  }
+
+  // if user sets 'No' for a DB/Schema and a sub schema/tables contains 'Blocked' permissions, warn them
+  // that we'll automatically upgrade the DB/Schema to 'Unrestricted' view access
+  const hasCreateQueryAccess = value !== DataPermissionValue.NO;
+  if (!hasCreateQueryAccess) {
+    return;
+  }
+
+  const hasChildWithBlockedPermission = hasPermissionValueInSubgraph(
     permissions,
     groupId,
     entityId,
+    database,
     DataPermission.VIEW_DATA,
+    DataPermissionValue.BLOCKED,
   );
 
-  if (
-    value === DataPermissionValue.QUERY_BUILDER_AND_NATIVE &&
-    nativePermission !== DataPermissionValue.QUERY_BUILDER_AND_NATIVE &&
-    PLUGIN_ADVANCED_PERMISSIONS.shouldShowViewDataColumn &&
-    ![
-      DataPermissionValue.UNRESTRICTED,
-      DataPermissionValue.IMPERSONATED,
-    ].includes(viewPermission)
-  ) {
+  if (hasChildWithBlockedPermission) {
+    const entityType = isSchemaEntityId(entityId) ? t`schema` : t`database`;
+
     return {
-      title: t`Allow native query editing?`,
-      message: t`This will also change this group's data access to Unrestricted for this database.`,
+      title: t`Upgrade “View Data” to “Unrestricted” for this ${entityType}?`,
+      message: t`Your “Granular” permissions for “View Data” contains “Blocked” which restricts “Create Queries” access. Would you like to upgrade the permissions for this ${entityType} to “Unrestricted” to allow for this change?`,
```

Comment:
```suggestion
      title: isSchemaEntity(entityId) ? t`Upgrade “View Data” to “Unrestricted” for this schema?` : t`Upgrade “View Data” to “Unrestricted” for this database?`,
      message: t`Your “Granular” permissions for “View Data” contains “Blocked” which restricts “Create Queries” access. ` + isSchemaEntity(entityId) ? t`Would you like to upgrade the permissions for this schema to “Unrestricted” to allow for this change?` : t`Would you like to upgrade the permissions for this database to “Unrestricted” to allow for this change?`
```

Reply:
I wrote this copy so it's sure to change ([slack thread](https://metaboat.slack.com/archives/C064EB1UE5P/p1723506272929749)). I can made the feedback when I go to make whatever copy changes are asked of me.